File: LeetCode/224-basic-calculator.py
###############################################################################################
# 某种程度上比ACwing的那道题简单，因为没有*、/法，但多了一个单目运算符和空格
###########
# 时间复杂度：O(n)
# 空间复杂度：O(n)，存放num和op的数组开销总和
###############################################################################################
class Solution:
    def calculate(self, s: str) -> int:
        stack_num = []
        stack_op = []
        
        priority = {'(':0,'+':1,'-':1,'--':2} # 栈内优先级, '--'表示单目减号
        
        def eval_():
            op = stack_op.pop()
            if op != '--':
                x1, x2 = stack_num.pop(), stack_num.pop()
                stack_num.append(x1 + x2 if op == '+' else x2 - x1)
            else:
                x = stack_num.pop()
                stack_num.append(-x)

        i = 0
        while i < len(s):
            if s[i].isdigit():
                num = 0
                while i<len(s) and s[i].isdigit():
                    num = num*10 + int(s[i])
                    i += 1
                stack_num.append(num)
                i -= 1
            elif s[i] == '(':
                stack_op.append(s[i])
            elif s[i] == ')':
                while stack_op[-1] != '(':
                    eval_()
                stack_op.pop() # 弹出左括号
            elif s[i] == '+' or s[i] == '-':
                if len(stack_op) and priority[stack_op[-1]] >= priority[s[i]]:
                    eval_()
                stack_op.append('--' if s[i] == '-' and (s[i-1] == '(' or not i) else s[i]) # 处理单目
            i += 1
        while len(stack_op):
            eval_()
        return stack_num[-1]

File: LeetCode/224-basic-calculator.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// hard, it turns out that my previous python solution has bugs but didn't been tested out last time
////////////////
// Time Complexity: O(n)
// Space Complexity: O(n)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    vector<int> stack_num;
    vector<char> stack_op;
    // 栈内优先级, '_' means negative
    unordered_map<char, int> priority = {{'(',0},{'+',1},{'-',1},{'_',2}};
    void eval_(){
        char op = stack_op.back();
        stack_op.pop_back();
        // not negative
        if (op != '_') {
            int x1 = stack_num.back();
            stack_num.pop_back();
            int x2 = stack_num.back();
            stack_num.pop_back();
            
            stack_num.push_back(op == '+'? x1 + x2 : x2 - x1);
        }
        // negative
        else {
            int x = stack_num.back();
            stack_num.pop_back();
            stack_num.push_back(-x);
        }
    }
    int calculate(string s) {
        char last_sign = ' ';
        for (int i=0 ; i<s.length() ; i++) {
            if (s[i] >= '0') {
                int num = 0;
                while (i<s.length() && s[i] >= '0') 
                    num = num*10 + (s[i++] - '0');
                stack_num.push_back(num);
                i --;
            }
            else if (s[i] == '(') {
                stack_op.push_back(s[i]);
            }
            else if (s[i] == ')') {
                // consume
                while (stack_op.back() != '(') eval_();
                // 弹出左括号
                stack_op.pop_back();
            }
            else if (s[i] == '+' || s[i] == '-') {
                if (stack_op.size() && priority[stack_op.back()] >= priority[s[i]]) {
                    eval_();
                }
                stack_op.push_back((s[i] == '-' && (last_sign == '(' || i==0)) ? '_' : s[i]);
                
            }
            if (s[i] != ' ') last_sign = s[i];
        }
        
        // consume left operations
        while (stack_op.size() > 0) eval_();
        
        return stack_num.back();
    }
};